lib/openstack_query/runners/server_runner.py

```diff
@@ -0,0 +1,104 @@
+
+from openstack_api.openstack_connection import OpenstackConnection
+from openstack_query.runners.query_runner import QueryRunner
+
+from exceptions.parse_query_error import ParseQueryError
+
+
+class ServerRunner(QueryRunner):
+    """
+    Runner class for openstack Server resource.
+    ServerRunner encapsulates running any openstacksdk Server commands
+    """
+
+    def _run_query(
+        self,
+        conn: OpenstackConnection,
+        filter_kwargs: Optional[Dict[str, str]] = None,
+        **kwargs
+    ) -> List[Server]:
+        """
+        This method runs the query by running openstacksdk commands
+
+        For QueryServer, this command gets all projects available and iteratively finds servers that belong to that project
+        :param conn: An OpenstackConnection object - used to connect to openstacksdk
+        :param filter_kwargs: An Optional set of filter kwargs to pass to conn.compute.servers()
+            to limit the servers being returned. - see https://docs.openstack.org/api-ref/compute/#list-servers
+
+        :param kwargs: a set of extra meta params that override default openstacksdk command. Current valid kwargs are:
+
+            'from_projects': Union[List[str], List[Project]] - takes a list of projects to run the query on
+
+            'from_subset': Union[List[str], List[Server]] - takes a list of servers to run the query on
+            (forces use of filter functions)
+
+        """
+        projects = []
+        if "from_projects" in kwargs:
+            if len(kwargs["from_projects"]) == 0:
+                raise ParseQueryError(
+                    "'from_project' meta kwarg given but no projects given to search in"
+                )
+            for project in kwargs["from_projects"]:
+                if not isinstance(project, Project):
+                    raise ParseQueryError(
+                        "'from_project' only accepts Openstack 'Project' objects"
+                    )
+                projects.append(project)
+        else:
+            for project in conn.identity.projects():
+                projects.append(project)
+
+        return self._run_query_on_projects(conn, projects, filter_kwargs)
+
+    def _run_query_on_projects(
+        self,
+        conn: OpenstackConnection,
+        projects: List[Project],
+        filter_kwargs: Optional[Dict[str, str]] = None,
+    ) -> List[Server]:
+        """
+        This method is a helper function that will run the query on a list of openstack projects given
+        :param conn: An OpenstackConnection object - used to connect to openstacksdk
+        :param projects: A list of openstacksdk projects to run query on
+        :param filter_kwargs: An Optional set of filter kwargs to pass to conn.compute.servers()
+        """
+        return [
+            server
+            for project in projects
+            for server in self._run_query_on_project(conn, project, filter_kwargs)
+        ]
```

Comment:
Do we not group by project?
I would argue the that `Dict[ProjectIdentifier, Server]` is more useful and people can always just do `servers = result.values()` at the other end

Reply:
I can do this for this method `_run_query_on_projects` - but overall I want `run_query` to return all servers not grouped by anything - I want the user to explicitly tell us to group by projects - via calling `group_by` 
e.g.: `QueryServer().group_by(Project).to_list()`

Reply:
I've changed this function to return a dictionary `{project_id: [servers]}` - then I flatten this into a list of servers in `run_query` - if we decide that we want to group by projects as default then I can change `run_query` but for now I left it as is - changing `run_query` will mean reworking `QueryOutput` a bit

Reply:
I'm more than happy with this, since it should also make it easier when we do a `groupBy` in the query layer later on too :+1: